Controllers/User.go

```diff
@@ -1,109 +1,129 @@
-package Controllers
-import(
+package controller
+
+// package name should start with small letters
+import (
 	"first-api/Models"
 	"fmt"
 	"net/http"
+
 	"github.com/gin-gonic/gin"
 )
 
-
 // get all users
 
-func GetUsers(c *gin.Context){
-	var user []Models.User
-	err := Models.GetAllUsers(&user);
-	if err!= nil{
-		fmt.Println(err.Error())
-		c.AbortWithStatus(http.StatusNotFound)
-	}else{
-		c.JSON(http.StatusOK,user)
-	}
-
+type UserStorer interface {
+	CreateUser(user *model.User) error
+	Validate(user model.User) error
+	GetAllUsers(users *[]model.User) error
+	GetUserByID(user *model.User, id string) error
+	UpdateUser(user *model.User, id string) error
+	DeleteUser(user *model.User, id string) error
 }
 
-func CreateUser(c *gin.Context){
-	var user Models.User
-	c.BindJSON(&user)
-	valErr := user.Validate();
-	if valErr != nil {
-		fmt.Println(valErr) 
-		c.JSON(http.StatusNotFound,valErr)
-		return 
-	}
-
-	err := Models.CreateUser(&user)
-	fmt.Println("user:",user);
-	if err!=nil {
-		fmt.Println(err.Error())
-		c.AbortWithStatus(http.StatusNotFound)
-	}else{
-		fmt.Println(user.Name)
-		c.JSON(http.StatusOK,user);
+func GetUserByIDController(store UserStorer) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		id := c.Params.ByName("id")
+		var user model.User
+
+		err := store.GetUserByID(&user, id)
+		if err != nil {
+			fmt.Println(err.Error())
+			c.AbortWithStatus(http.StatusNotFound)
```

Comment:
this is not correct. 

if err != nil , there could be possibility that database connection is blocked or may be some other reasons.